### PR TITLE
Seniors and Staff Exempt from Anti-Raid

### PR DESCRIPTION
**Pull request by austinmccalley**
Staff and seniors are now exempt from anti-raid

### CODE REVIEW

File: src/processes/antiraidListener.ts [hunk 1]
```diff
@@ -16,29 +16,45 @@ export const antiraidListener: ProcessInterface = {
...
+
+            // Staff role is not used as of now
+            const staff =
+                userRoles.find(role => role.id === staffRole) != undefined
+            const senior =
+                userRoles.find(role => role.id === seniorRole) != undefined
+
+            // NOTE: Cant combine the two roles else theres weird operator logic
+            if (!senior) {
+                if (!staff) {
```
Comment (lostdesign) [inline src/processes/antiraidListener.ts:42]:
```suggestion
if (!senior || !staff) {
```
?

**Comment by redxtech** [comment]
Try using `.some` on an array of .staff & senior.

`['staff', 'senior'].some(r => user.roles.includes(r))`

Not valid code but got the right idea.

File: src/processes/antiraidListener.ts [hunk 2]
```diff
@@ -16,29 +16,45 @@ export const antiraidListener: ProcessInterface = {
...
-                mentionedUsers + mentionedRoles > 2
-            ) {
-                lastKnownTimestamp
-                    ? checkOffender(message, lastKnownTimestamp)
-                    : addOffenderToList(message)
+            const userRoles = message.member.roles.cache
+
+            // Staff role is not used as of now
+            const staff =
+                userRoles.find(role => role.id === staffRole) != undefined
```
Comment (waffln) [inline src/processes/antiraidListener.ts:36]:
```js
const isStaff = userRoleIds.includes(staffRoleId)
```

File: src/processes/antiraidListener.ts [hunk 3]
```diff
@@ -16,29 +16,45 @@ export const antiraidListener: ProcessInterface = {
...
-
-            if (
-                mentionedRoles > 1 ||
-                mentionedUsers > 2 ||
-                mentionedUsers + mentionedRoles > 2
-            ) {
-                lastKnownTimestamp
-                    ? checkOffender(message, lastKnownTimestamp)
-                    : addOffenderToList(message)
+            const userRoles = message.member.roles.cache
```
Comment (waffln) [inline src/processes/antiraidListener.ts:32]:
```js
const userRoleIds = message.member.roles.cache.map(r => r.id)
```

File: src/processes/antiraidListener.ts [hunk 4]
```diff
@@ -16,29 +16,45 @@ export const antiraidListener: ProcessInterface = {
         const mentions = new Map()
         const mutedRole = config.roleIDs.muted
 
-        // TODO: fix staffRole to mention properly, cannot be string
-        //const staffRole = config.roleIDs.staff
-        const staffRole = 662405675107876864
+        // NOTE: Pull from config, same role is achieved
+        const staffRole = config.roleIDs.staff
```
Comment (waffln) [inline src/processes/antiraidListener.ts:20]:
```js
const staffRoleId = config.roleIDs.staff // better name
```

File: src/processes/antiraidListener.ts (hunk 1 above)
Comment (waffln) [reply src/processes/antiraidListener.ts:42]: Also
```js
if (isSenior || isStaff) {
  return
}
```

Comment (austinmccalley) [reply src/processes/antiraidListener.ts:42]: Doesn't work due to if there is a staff member that isn't a senior then it prunes the message.
So say that ``Frank`` has the roles ``Staff`` but not ``Senior`` then the if statement evaluates to ``true || false`` resulting in ``true`` going through the statement. This is why it needs to be seperated out into two different if statements.

File: src/processes/antiraidListener.ts (hunk 2 above)
Comment (austinmccalley) [reply src/processes/antiraidListener.ts:36]: Type errors

File: src/processes/antiraidListener.ts (hunk 1 above)
Comment (lostdesign) [reply src/processes/antiraidListener.ts:42]: @austinmccalley Check what @waffln said? Thats the solution.

Comment (waffln) [reply src/processes/antiraidListener.ts:42]: @austinmccalley If the user is senior, or he is staff, return (therefore aborting raid-check).

File: config.example.ts [hunk 5]
```diff
@@ -9,6 +9,8 @@ export default <UserConfigInterface>{
     dbFile: '',
     autoBan: true,
     autoBanThreshold: 3,
+    antiRaidMembers: 2,
```
Comment (lostdesign) [inline config.example.ts:12]:
```suggestion
    antiRaidMembers: 3,
```

File: src/processes/antiraidListener.ts [hunk 6]
```diff
@@ -16,29 +16,46 @@ export const antiraidListener: ProcessInterface = {
         const mentions = new Map()
         const mutedRole = config.roleIDs.muted
 
-        // TODO: fix staffRole to mention properly, cannot be string
-        //const staffRole = config.roleIDs.staff
-        const staffRole = 662405675107876864
+        // NOTE: Pull from config, same role is achieved
+        const staffRoleId = config.roleIDs.staff
+        // const staffRole = 662405675107876864
```
Comment (lostdesign) [inline src/processes/antiraidListener.ts:21]:
```suggestion

```

File: src/processes/antiraidListener.ts [hunk 7]
```diff
@@ -16,29 +16,46 @@ export const antiraidListener: ProcessInterface = {
...
-                mentionedRoles > 1 ||
-                mentionedUsers > 2 ||
-                mentionedUsers + mentionedRoles > 2
-            ) {
-                lastKnownTimestamp
-                    ? checkOffender(message, lastKnownTimestamp)
-                    : addOffenderToList(message)
+            const userRoleIds = message.member.roles.cache.map(r => r.id)
+
+            // Staff role is not used as of now
```
Comment (lostdesign) [inline src/processes/antiraidListener.ts:38]: Mhh?

File: src/processes/antiraidListener.ts [hunk 8]
```diff
@@ -16,29 +16,46 @@ export const antiraidListener: ProcessInterface = {
...
+            const isStaff = userRoleIds.find(r => r === staffRoleId)
+            const isSenior = userRoleIds.find(r => r === seniorRoleId)
+
+            if (isStaff || isSenior) {
+                return
+            } else {
```
Comment (lostdesign) [inline src/processes/antiraidListener.ts:44]: Guess else is not really necessary? Any reason to keep this?

Comment (austinmccalley) [reply src/processes/antiraidListener.ts:43]: Just styling and better readability.

File: src/processes/antiraidListener.ts (hunk 7 above)
Comment (austinmccalley) [reply src/processes/antiraidListener.ts:37]: Get a list of all the users roles.

Comment (lostdesign) [reply src/processes/antiraidListener.ts:37]: I meant the comment in Line 37 :D I dont think its neccessary anymore.

File: src/processes/antiraidListener.ts [hunk 9]
```diff
@@ -16,29 +16,44 @@ export const antiraidListener: ProcessInterface = {
...
-                    ? checkOffender(message, lastKnownTimestamp)
-                    : addOffenderToList(message)
+            const userRoleIds = message.member.roles.cache.map(r => r.id)
+
+            const isStaff = userRoleIds.find(r => r === staffRoleId)
+            const isSenior = userRoleIds.find(r => r === seniorRoleId)
+
+            if (isStaff || isSenior) {
+                return
+            } else {
```
Comment (waffln) [inline src/processes/antiraidListener.ts:42]: No else! return -> end of function

File: src/processes/antiraidListener.ts [hunk 10]
```diff
@@ -16,29 +16,44 @@ export const antiraidListener: ProcessInterface = {
...
+                    users: { size: mentionedUsers },
+                    roles: { size: mentionedRoles }
+                } = message.mentions
+
+                if (
+                    mentionedRoles > maxMentionedRoles ||
+                    mentionedUsers > maxMentionedMemebers ||
+                    mentionedUsers + mentionedRoles > maxMentionedEverything
+                ) {
+                    lastKnownTimestamp
```
Comment (waffln) [inline src/processes/antiraidListener.ts:53]: Specific comparison like === null

**Pull request merged**